### ISSUE TITLE
Cannot provision multi-node cluster when using Cilium with AWS ENI enabled

### ISSUE THREAD

**Issue by riuvshyn** [rancher/rke2#3108]
**Environmental Info:**
RKE2 Version: **v1.22.10+rke2r2**

Node(s) CPU architecture, OS, and Version: **AWS c5.xlarge**

Cluster Configuration:
3 control-plane nodes, 3 workers

**Describe the bug:**
Some `kubelet` args can't be set in `KubeletConfiguration`, example is `--node-ip=` this can't be set in `rke_config` as every node must use it's own IP and there is no way pass NODE-IP as kubelet arg.
Perhaps there could be some well known ENV var like `RKE2_EXTRA_KUBELET_ARGS` that would allow to pickup these args on the host, example: `RKE2_EXTRA_KUBELET_ARGS=node-ip=xxx.xxx.xxx.xxx,--provider-id=aws:///eu-central-1c/i-xxxxxxxxxxxxxxx`

**Steps To Reproduce:**
Try to configure `--node-ip=` kubelet arg

**Expected behavior:**
There must be a way to set any `kubelet` argument

**Actual behavior:**
Unable to set dynamic `kubelet` args

**Additional context / logs:**
I've tried to pass something like 
```
rkeConfig:
...
  machineGlobalConfig:
    kubelet-arg:
      - node-ip=$(ec2metadata --local-ipv4)
      OR
      - node-ip=${KUBELET_NODE_IP}
```
but nothing helps
kubelet logs:
```
server.go:1127] "Could not parse --node-ip ignoring" IP="$(ec2metadata --local-ipv4)"
```
OR
```
server.go:1127] "Could not parse --node-ip ignoring" IP="${RKE2_NODE_IP}"
```

**Comment by riuvshyn** [comment]
This also makes impossible to use Cilium CNI with AWS ENI

**Comment by riuvshyn** [comment]
@brandond hello, thanks for looking in to this. 
Yes, I know that there is top level `--node-ip` how ever it is still cluster configuration which I can't set up beforehand as I don't know what IPs are going to be there. 
If https://github.com/rancher/system-agent/blob/main/install.sh was able to accept `--node-ip` as it does for `--node-name` then I would be able to pass it from user-data
So I know the IP only when instance is running and the only way where i can pass things  is registration command
like this: 
```
curl --insecure -fL https://rancher.test.com/system-agent-install.sh | sudo  sh -s - --server https://rancher.test.com --label 'cattle.io/os=linux' --token xxx --ca-checksum xxx --etcd --controlplane  --node-name $(ec2metadata --local-hostname)
```

**Comment by riuvshyn** [comment]
oh, that's interesting I will try this. It wasn't clear from the docs that this is even possible.
Thanks, I will try it and then will reply if that worked or not. 🙇🏽

**Comment by riuvshyn** [comment]
@brandond just tried that, it doesn't look that `node-ip: xxx.xxx.xxx.xxx` from `/etc/rancher/rke2/config.yaml.d/99-node-ip.yaml` (created before rke2) was actually passed to the `kubelet`, I do not see that in args on logs, node `status.addresses` still getting POD IPs and other control plan nodes got wrong server IP.

**Comment by riuvshyn** [comment]
however this:
`/etc/rancher/rke2/config.yaml.d/99-node-ip.yaml`
```
{
  "kubelet-arg": [
    "node-ip=172.29.195.10"
  ]
}
```
seems to be working!

**Comment by riuvshyn** [comment]
it looks like it is also supporting json, as `/etc/rancher/rke2/config.yaml.d/50-rancher.yaml` which was provisioned by installer I guess is in json format :)

**Comment by brandond** [comment]
As of YAML 1.2, all valid JSON is also valid YAML.

**Comment by riuvshyn** [comment]
hmm... now I see `kubelet` actually got `--node-ip` set correctly however 
`kubectl get nodes -oyaml`  still shows
```
  status:
    addresses:
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: ip-xxx-xxx-xxx-xxx.eu-central-1.compute.internal
      type: InternalDNS
    - address: ip-xxx-xxx-xxx-xxx.eu-central-1.compute.internal
      type: Hostname
```

That is weird, I thought setting up `node-ip` will leave only node-ip there in addresses...

**Comment by riuvshyn** [comment]
ohh... I think it dropped all other kubelet-args I've set via TF in `rke_config`

**Comment by brandond** [comment]
There's more poorly documented or undocumented behavior here, but for flags that take a list, you can append to the existing list instead of replacing it by ending the key with `+`:
```yaml
kubelet-arg+: ["node-ip=172.29.195.10"]
```

**Comment by riuvshyn** [comment]
Is there any difference how bootstrap control plane node starts up first time and others?
other control plane nodes if I fix server address manually don't have POD IPs in `status.addresses`
So I was thinking maybe first time kubelet is launched with some other set of args and then restarts with what in configuration.

**Comment by brandond** [comment]
I'm confused, are you talking about the pods missing IPs, or the nodes missing IPs?

**Comment by brandond** [comment]
Are you trying to use the [AWS VPC CNI](https://github.com/aws/amazon-vpc-cni-k8s) alongside Cilium? I'm unclear on why your interfaces have multiple IPs. I suspect that this isn't a supported configuration.

**Comment by brandond** [comment]
I suspect something else is going on here. Are these clean nodes, or are they left over from a previous installation? Can you on a fresh instance create a single-node cluster with only Cilium, and once it comes up, provide the output of `kubectl get node -o yaml`?

**Comment by riuvshyn** [comment]
here is a node output of a new cluster with cilium cni + enabled aws eni

```
kubectl get nodes -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      alpha.kubernetes.io/provided-node-ip: xxx.xxx.xxx.xxx
      etcd.rke2.cattle.io/node-address: xxx.xxx.xxx.xxx
      etcd.rke2.cattle.io/node-name: -bc427852
      io.cilium.network.ipv4-cilium-host: xxx.xxx.xxx.xxx
      io.cilium.network.ipv4-health-ip: xxx.xxx.xxx.xxx
      io.cilium.network.ipv4-pod-cidr: 100.64.0.0/24
      management.cattle.io/pod-limits: '{"cpu":"220m","memory":"720Mi"}'
      management.cattle.io/pod-requests: '{"cpu":"1320m","memory":"820Mi","pods":"14"}'
      node.alpha.kubernetes.io/ttl: "0"
      rke2.io/encryption-config-hash: start-94471b65897ba37dfe85c4a21ed8b64d9d36f79e138cfdb406d9b60fa9c316bd
      rke2.io/node-args: '["server","--agent-token","********","--cloud-provider-name","external","--cluster-cidr","100.64.0.0/13","--cluster-dns","169.254.20.10","--cluster-domain","cluster.local","--cni","cilium","--disable","rke2-ingress-nginx","--disable","rke2-metrics-server","--disable","rke2-coredns","--disable","rke2-canal","--disable-cloud-controller","true","--etcd-s3","true","--etcd-s3-bucket","etcd-backups-rancher-euc1-test-test01","--etcd-s3-endpoint","s3.eu-central-1.amazonaws.com","--etcd-s3-folder","etcd","--etcd-snapshot-retention","12","--etcd-snapshot-schedule-cron","0
        */6 * * *","--kube-apiserver-arg","allow-privileged=true","--kube-apiserver-arg","anonymous-auth=false","--kube-apiserver-arg","feature-gates=CustomCPUCFSQuotaPeriod=true","--kube-apiserver-arg","api-audiences=https://rancher-euc1-test-test01-oidc.s3.eu-central-1.amazonaws.com,https://kubernetes.default.svc.cluster.local,rke2","--kube-apiserver-arg","audit-log-maxage=90","--kube-apiserver-arg","audit-log-maxbackup=10","--kube-apiserver-arg","audit-log-maxsize=500","--kube-apiserver-arg","audit-log-path=/var/log/k8s-audit.log","--kube-apiserver-arg","audit-policy-file=/etc/kubernetes/audit-policy.yaml","--kube-apiserver-arg","authorization-mode=Node,RBAC","--kube-apiserver-arg","bind-address=0.0.0.0","--kube-apiserver-arg","enable-admission-plugins=PodSecurityPolicy,NodeRestriction","--kube-apiserver-arg","event-ttl=1h","--kube-apiserver-arg","kubelet-preferred-address-types=InternalIP,Hostname,ExternalIP","--kube-apiserver-arg","profiling=false","--kube-apiserver-arg","request-timeout=60s","--kube-apiserver-arg","runtime-config=api/all=true","--kube-apiserver-arg","service-account-key-file=/etc/kubernetes/service-account.pub","--kube-apiserver-arg","service-account-lookup=true","--kube-apiserver-arg","service-account-issuer=https://rancher-euc1-test-test01-oidc.s3.eu-central-1.amazonaws.com","--kube-apiserver-arg","service-account-signing-key-file=/etc/kubernetes/service-account.key","--kube-apiserver-arg","service-node-port-range=30000-32767","--kube-apiserver-arg","shutdown-delay-duration=60s","--kube-apiserver-arg","token-auth-file=/etc/kubernetes/static-auth-tokens.csv","--kube-apiserver-arg","tls-min-version=VersionTLS12","--kube-apiserver-arg","tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384","--kube-apiserver-arg","tls-min-version=VersionTLS12","--kube-apiserver-arg","v=2","--kube-apiserver-extra-mount","/etc/kubernetes:/etc/kubernetes:ro","--kube-controller-manager-arg","allocate-node-cidrs=true","--kube-controller-manager-arg","attach-detach-reconcile-sync-period=1m0s","--kube-controller-manager-arg","bind-address=0.0.0.0","--kube-controller-manager-arg","configure-cloud-routes=false","--kube-controller-manager-arg","feature-gates=CustomCPUCFSQuotaPeriod=true","--kube-controller-manager-arg","leader-elect=true","--kube-controller-manager-arg","node-monitor-grace-period=2m","--kube-controller-manager-arg","pod-eviction-timeout=220s","--kube-controller-manager-arg","profiling=false","--kube-controller-manager-arg","service-account-private-key-file=/etc/kubernetes/service-account.key","--kube-controller-manager-arg","use-service-account-credentials=true","--kube-controller-manager-arg","terminated-pod-gc-threshold=12500","--kube-controller-manager-arg","tls-min-version=VersionTLS12","--kube-controller-manager-arg","tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384","--kube-controller-manager-arg","tls-min-version=VersionTLS12","--kube-controller-manager-arg","cert-dir=/var/lib/rancher/rke2/server/tls/kube-controller-manager","--kube-controller-manager-arg","secure-port=10257","--kube-controller-manager-extra-mount","/etc/kubernetes:/etc/kubernetes:ro","--kube-controller-manager-extra-mount","/var/lib/rancher/rke2/server/tls/kube-controller-manager:/var/lib/rancher/rke2/server/tls/kube-controller-manager","--kube-proxy-arg","conntrack-max-per-core=131072","--kube-proxy-arg","conntrack-tcp-timeout-close-wait=0s","--kube-proxy-arg","metrics-bind-address=0.0.0.0","--kube-proxy-arg","proxy-mode=iptables","--kube-scheduler-arg","bind-address=0.0.0.0","--kube-scheduler-arg","port=0","--kube-scheduler-arg","secure-port=10259","--kube-scheduler-arg","profiling=false","--kube-scheduler-arg","leader-elect=true","--kube-scheduler-arg","tls-min-version=VersionTLS12","--kube-scheduler-arg","v=2","--kube-scheduler-arg","tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384","--kube-scheduler-arg","tls-min-version=VersionTLS12","--kube-scheduler-arg","cert-dir=/var/lib/rancher/rke2/server/tls/kube-scheduler","--kube-scheduler-extra-mount","/var/lib/rancher/rke2/server/tls/kube-scheduler:/var/lib/rancher/rke2/server/tls/kube-scheduler","--kubelet-arg","feature-gates=CustomCPUCFSQuotaPeriod=true","--kubelet-arg","config=/etc/kubernetes/kubelet.yaml","--kubelet-arg","exit-on-lock-contention=true","--kubelet-arg","lock-file=/var/run/lock/kubelet.lock","--kubelet-arg","pod-infra-container-image=docker-k8s-gcr-io.test.com/pause:3.1","--kubelet-arg","register-node=true","--kubelet-arg","tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384","--kubelet-arg","tls-min-version=VersionTLS12","--kubelet-arg","v=2","--kubelet-arg","node-ip=xxx.xxx.xxx.xxx","--node-label","cattle.io/os=linux","--node-label","rke.cattle.io/machine=*********","--node-name","ip-xxx.xxx.xxx.xxx.eu-central-1.compute.internal","--node-taint","node-role.kubernetes.io/control-plane:NoSchedule","--node-taint","node-role.kubernetes.io/etcd:NoExecute","--private-registry","/etc/rancher/rke2/registries.yaml","--service-cidr","100.64.0.0/13","--system-default-registry","docker.test.com","--token","********","--node-ip","xxx.xxx.xxx.xxx"]'
      rke2.io/node-config-hash: SCC2HVXWAFCI26JSK4UT553CJADE5ULMHCWHAFEGB3GWEWZ4C47A====
      rke2.io/node-env: '{}'
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: "2022-06-27T22:10:58Z"
    finalizers:
    - wrangler.cattle.io/node
    - wrangler.cattle.io/managed-etcd-controller
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/instance-type: c5.xlarge
      beta.kubernetes.io/os: linux
      cattle.io/os: linux
      egress.rke2.io/cluster: "true"
      failure-domain.beta.kubernetes.io/region: eu-central-1
      failure-domain.beta.kubernetes.io/zone: eu-central-1c
      kubernetes.io/arch: amd64
      kubernetes.io/hostname: ip-xxx.xxx.xxx.xxx.eu-central-1.compute.internal
      kubernetes.io/os: linux
      node-role.kubernetes.io/control-plane: "true"
      node-role.kubernetes.io/etcd: "true"
      node-role.kubernetes.io/master: "true"
      node.kubernetes.io/instance-type: c5.xlarge
      rke.cattle.io/machine: *********
      topology.kubernetes.io/region: eu-central-1
      topology.kubernetes.io/zone: eu-central-1c
    name: ip-xxx.xxx.xxx.xxx.eu-central-1.compute.internal
    resourceVersion: "1457"
    uid: 6ccbf8c5-6581-43de-9b58-7ae0db7e5c67
  spec:
    podCIDR: 100.64.0.0/24
    podCIDRs:
    - 100.64.0.0/24
    taints:
    - effect: NoSchedule
      key: node-role.kubernetes.io/control-plane
    - effect: NoExecute
      key: node-role.kubernetes.io/etcd
  status:
    addresses:
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: xxx.xxx.xxx.xxx
      type: InternalIP
    - address: ip-xxx.xxx.xxx.xxx.eu-central-1.compute.internal
      type: InternalDNS
    - address: ip-xxx.xxx.xxx.xxx.eu-central-1.compute.internal
      type: Hostname
    allocatable:
      cpu: "3"
      ephemeral-storage: "49255941901"
      hugepages-1Gi: "0"
      hugepages-2Mi: "0"
      memory: 6764268Ki
      pods: "110"
    capacity:
      cpu: "4"
      ephemeral-storage: 50633164Ki
      hugepages-1Gi: "0"
      hugepages-2Mi: "0"
      memory: 7788268Ki
      pods: "110"
    conditions:
    - lastHeartbeatTime: "2022-06-27T22:12:15Z"
      lastTransitionTime: "2022-06-27T22:12:15Z"
      message: Cilium is running on this node
      reason: CiliumIsUp
      status: "False"
      type: NetworkUnavailable
    - lastHeartbeatTime: "2022-06-27T22:12:48Z"
      lastTransitionTime: "2022-06-27T22:10:57Z"
      message: kubelet has sufficient memory available
      reason: KubeletHasSufficientMemory
      status: "False"
      type: MemoryPressure
    - lastHeartbeatTime: "2022-06-27T22:12:48Z"
      lastTransitionTime: "2022-06-27T22:10:57Z"
      message: kubelet has no disk pressure
      reason: KubeletHasNoDiskPressure
      status: "False"
      type: DiskPressure
    - lastHeartbeatTime: "2022-06-27T22:12:48Z"
      lastTransitionTime: "2022-06-27T22:10:57Z"
      message: kubelet has sufficient PID available
      reason: KubeletHasSufficientPID
      status: "False"
      type: PIDPressure
    - lastHeartbeatTime: "2022-06-27T22:12:48Z"
      lastTransitionTime: "2022-06-27T22:12:08Z"
      message: kubelet is posting ready status. AppArmor enabled
      reason: KubeletReady
      status: "True"
      type: Ready
    daemonEndpoints:
      kubeletEndpoint:
        Port: 10250
    images:
    - names:
      - docker.test.com/rancher/rancher-agent@sha256:eb4c1eaae3bd0c0881874f65f20a847a40de75b5a3399d2c7979cb0e78a3c827
      - docker.test.com/rancher/rancher-agent:v2.6.5
      sizeBytes: 452543995
    - names:
      - docker.test.com/rancher/hardened-kubernetes@sha256:643d6c2e6e2a9cb423580c04e8b61cb516f74ed7f8c834784417e521697bca57
      - docker.test.com/rancher/hardened-kubernetes:v1.22.10-rke2r2-build20220608
      sizeBytes: 220511597
    - names:
      - docker.test.com/rancher/mirrored-cilium-cilium@sha256:fe2a5f0b9a10c423b5a1b25f1314490f099a240a7f0820b827ef3ffdaad1b6a6
      - docker.test.com/rancher/mirrored-cilium-cilium:v1.11.5
      sizeBytes: 158523134
    - names:
      - docker.test.com/rancher/klipper-helm@sha256:7d0d23c914fadde514e42d81355abe2fb8c6cb848717913db055e2034c17b305
      - docker.test.com/rancher/klipper-helm:v0.7.1-build20220407
      sizeBytes: 82974686
    - names:
      - docker.test.com/rancher/hardened-etcd@sha256:fecba6eb3e3e356126590486967fec13d780b55aee82706af2fe58d32c718a08
      - docker.test.com/rancher/hardened-etcd:v3.5.4-k3s1-build20220504
      sizeBytes: 49072207
    - names:
      - docker-k8s-gcr-io.test.com/k8s-dns-node-cache@sha256:8e765f63b3a5b4832c484b4397f4932bd607713ec2bb3e639118bc164ab4a958
      - docker-k8s-gcr-io.test.com/k8s-dns-node-cache:1.15.12
      sizeBytes: 42055085
    - names:
      - docker.test.com/rancher/mirrored-cilium-operator-aws@sha256:bf7c66163b277d90a27032f53adf33ee6d3e7fafb098f4aa8754db9da89c5baa
      - docker.test.com/rancher/mirrored-cilium-operator-aws:v1.11.5
      sizeBytes: 20213790
    - names:
      - docker-k8s-gcr-io.test.com/provider-aws/cloud-controller-manager@sha256:3a607f27e0da4663818df58b4624882a6746602a3eee2e14c4a5345d2edb6104
      - docker-k8s-gcr-io.test.com/provider-aws/cloud-controller-manager:v1.22.0-alpha.0
      sizeBytes: 16241849
    - names:
      - docker-k8s-gcr-io.test.com/cluster-proportional-autoscaler-amd64@sha256:be8875e5584750b7a490244ee56a121a714aa3d124164a5090cd8b3570c5650f
      - docker-k8s-gcr-io.test.com/cluster-proportional-autoscaler-amd64:1.8.1
      sizeBytes: 15208262
    - names:
      - docker.test.com/company/kubernetes/coredns@sha256:8916c89e1538ea3941b58847e448a2c6d940c01b8e716b20423d2d8b189d3972
      - docker.test.com/company/kubernetes/coredns:1.8.6-211013-101104-05c4688a
      sizeBytes: 13583131
    - names:
      - docker.test.com/rancher/mirrored-cilium-startup-script@sha256:8fea347d29f3b9ac8bfba6023c73bcbc9ee89e9553e53c6bd6b52137a1813073
      - docker.test.com/rancher/mirrored-cilium-startup-script:62bfbe88c17778aad7bef9fa57ff9e2d4a9ba0d8
      sizeBytes: 5252537
    - names:
      - docker.test.com/container-static-tools@sha256:c7851b975e53dabec0cd16061287aa2fdee42315175d62d7a0f74db062013bd6
      - docker.test.com/container-static-tools:1.191206.164824-6ffbe7e
      sizeBytes: 2870347
    - names:
      - docker.test.com/rancher/pause@sha256:74bf6fc6be13c4ec53a86a5acf9fdbc6787b176db0693659ad6ac89f115e182c
      - docker.test.com/rancher/pause:3.6
      sizeBytes: 297944
    nodeInfo:
      architecture: amd64
      bootID: 7651d80b-7c33-46d9-90ca-2d9e01a3ce87
      containerRuntimeVersion: containerd://1.5.13-k3s1
      kernelVersion: 5.4.0-1080-aws
      kubeProxyVersion: v1.22.10+rke2r2
      kubeletVersion: v1.22.10+rke2r2
      machineID: ec2e402685f3165f18116e896082b509
      operatingSystem: linux
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by riuvshyn** [comment]
hmmm wondering if this is something related to https://github.com/kubernetes/kubernetes/pull/107750

**Comment by riuvshyn** [comment]
steps to provision with Cilium:
1. set `cni: cilium` in `rke_config.machine_global_config`
2. enable aws eni (passed as `additional_manifest`):
```
---
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-cilium
  namespace: kube-system
spec:
  valuesContent: |-
    eni:
      enabled: true
    ipam:
      mode: eni
```

additionally I have `cloud-provider-name: external` and `disable-cloud-controller: true`
that's pretty much it...

**Comment by brandond** [comment]
@manuelbuil @rancher-max is this something we've tested?

@Oats87 I suspect that the comment at https://github.com/rancher/rke2/issues/3108#issuecomment-1167887378 is correct, and Rancher is using the **last** IP as the server IP for other nodes to join, instead of the first. Although I am curious why the kubelet is adding the ENI secondary IPs as addresses at all; that seems unexpected.

**Comment by rancher-max** [comment]
> @manuelbuil @rancher-max Cilium's ENI mode something we've tested?

Not extensively. However, I did see it work in one instance (ref: https://github.com/rancher/rke2/issues/2859#issuecomment-1125347260). As mentioned above by @riuvshyn, setting the cloud-provider to explicitly use AWS should allow that. I'd be curious if you did want to disable the cloud provider and provide the out-of-tree aws cloud provider instead. I'd expect it to work with that as well, but that has not been tested at all.

**Comment by riuvshyn** [comment]
the issue was actually bug in kubelet which was fixed in 1.24.x, closing this.
